File: OO_design_patterns_training_2022-11/Solutions/C#/Exercise/XmlReporter.cs
using System;
using System.Collections.Generic;

namespace TestApp1
{
    public class XmlReporter : Reporter
    {
        protected override void CreateDataFile(List<string> data)
        {
            Console.WriteLine("Data written as Xml file:");
            foreach (var item in data)
            {
                Console.WriteLine(item);
            }
        }

        protected override void CreateFooterFile()
        {
            Console.WriteLine($"Wrote an imaginary footer as an Xml file to {OutputDir}");
        }

        protected override void CreateHeaderFile()
        {
            Console.WriteLine($"Wrote an imaginary header as an Xml file to {OutputDir}");
        }
    }
}

File: OO_design_patterns_training_2022-11/Solutions/C++/Exercise/src/XmlReporter.cpp
#include "XmlReporter.h"

#include <iostream>

void XmlReporter::CreateDataFile(const std::vector<std::string>& data) const
{
    std::cout << "Data written as XML file:" << std::endl;
    for (const auto& item : data)
    {
        std::cout << item << std::endl;
    }
}

void XmlReporter::CreateFooterFile() const
{
    std::cout << "Wrote an imaginary footer as an XML file to " << OutputDir() << std::endl;
}

void XmlReporter::CreateHeaderFile() const
{
    std::cout << "Wrote an imaginary header as an XML file to " << OutputDir() << std::endl;
}
